pkg/apis/topology/v1alpha2/helper/attribute/attribute.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package attribute
+
+import (
+	"github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha2"
+)
+
+// Find looks for attribute by its name in the provided AtributeList.
+// If found, returns the value of the attribute and an explicit boolean value equals true
+// Otherwise, returns the zero value and explicit boolean value equals false
+// In case there are duplicates in the attribute list, returns the first attribute from the left
+// found in iteration order, iterating on the attribute list treating it as a golang slice.
+func Find(attributes v1alpha2.AttributeList, name string) (string, bool) {
+	attr := findByName(attributes, name)
+	if attr == nil {
+		return "", false
+	}
+	return attr.Value, true
+}
+
+// Update updates or adds an attribute to the provided AttributeList, preserving the existing order.
+// The attribute name must be non-zero, the value can be zero.
+// If the attribute is already present in the AttributeList, its value is overridden with the given one.
+// If the attribute was not present in the AttributeList, adds a new attribute minimizing the changes
+// to the attribute list.
+// Returns the updated AttributeList and true if an attribute was created, false if modified in place.
+func Update(attributes v1alpha2.AttributeList, name, value string) (v1alpha2.AttributeList, bool) {
```

Comment:
I am wondering if passing `v1alpha2.Attribute` instead of `name, value string` wouldn't be more straightforward.

Reply:
and do we really care that much if the Attribute was created or modified in place?

Reply:
What do you think about

```
func (a *v1alpha2.AttributeList) Update( name, value string) bool
```

Reply:
that would work fine for in-place updates. But if the function needs to add a new element, it would need to return the mutated `AttributeList` (much like the builtin `append`). I reckon the name can be misleading - it's more like `UpdateOrCreate` or `Upsert` (thinking about SQL world), but all the alternatives I could think of are more awkward, first and foremost the two I mentioned here.

Reply:
> that would work fine for in-place updates. But if the function needs to add a new element, it would need to return the mutated AttributeList (much like the builtin append). I reckon the name can be misleading_

But if you have a pointer receiver it can append to the existing object. I'm not sure what would be the "right" way but it is possible to do real update

Reply:
that's a good point. I'll investigate a bit about how the API could look like and which form is more widespread in the community, or if there are any standard/recommendation

Reply:
FWIW I think the api could look something like

```golang
func (a *AttributeList) Update(name, value string)

func (a *AttributeList) Merge(new *AttributeList)

func (a *AttributeList) Get(name string) (AttributeInfo, bool)

func (a *AttributeList) Delete(name string) bool


```

Reply:
I'm reluctant to attach behaviour to API objects directly. I'll check what the standards/practices/recommendations are in the k8s ecosystem and update here accordingly.

Reply:
That is a fair point 👍 I remember that there are some helpers like this around in K8s (even in the core api?) but these might be some legacy stuff which is not encouraged anymore.

Reply:
As next step I'm going to implement the API proposed by @marquiz (with test and docs ofc). I'll do the aforementioned investigation in parallel. Once everything's ready we can have a final round of review. The functions will either have `AttributeList` as receiver or be free functions in the `helper/attributes` package like the current code does.